In [108]:
import pandas as pd
import numpy as np

## Vorbereitung Landtagswahl 2014

## Achtung: Wir verwenden für alle Zahlen int64!!!

Müssen wir am Ende vor der Ausspielung uwandeln. 

In [109]:
df_lw_2014 = pd.read_excel("./Rohdaten/LW_2014_original.xlsx", sheet_name="LW14_Ergebnisse_GE_TG")

In [110]:
# für Leipzig, Dresden, Chemnitz und Zwickau zusammengerechnete Werte pro Stadt ausspielen
# ist wichtig für die Karte, damit wir die Kommunen einheitlich färben können. 

#Leipzig
le = df_lw_2014[df_lw_2014["Ortname"].str.contains("Leipzig")]
leipzig = le.sum()
leipzig_series = pd.Series(leipzig)
leipzig_series["WK-Nr"] = 2733
leipzig_series["WK-Name"] = "Leipzig 1-7"
leipzig_series["AGS"] = 14713000
#Lösche die einzelnen Wahlkreise aus der Datei, da diese ansonsten falsche Gesamtzahlen aufweist... 
df_lw_2014 = df_lw_2014[~df_lw_2014["Ortname"].str.contains("Leipzig")]


In [111]:
df_lw_2014.shape

(439, 44)

In [112]:
#Dresden
dr = df_lw_2014[df_lw_2014["Ortname"].str.contains("Dresden")]
dresden = dr.sum() 
dresden_series = pd.Series(dresden)
dresden_series["WK-Nr"] = 4147
dresden_series["WK-Name"] = "Dresden 1-7"
dresden_series["AGS"] = 14612000
df_lw_2014 = df_lw_2014[~df_lw_2014["Ortname"].str.contains("Dresden")]

In [113]:
# Zwickau
zw = df_lw_2014[df_lw_2014["Ortname"].str.contains("Zwickau")]
zwickau = zw.sum() 
zwickau_series = pd.Series(zwickau)
zwickau_series["WK-Nr"] = 6677
zwickau_series["WK-Name"] = "Zwickau 2-3"
zwickau_series["AGS"] = 14524330
df_lw_2014 = df_lw_2014[~df_lw_2014["Ortname"].str.contains("Zwickau")]

In [114]:
# Chemnitz
ch = df_lw_2014[df_lw_2014["Ortname"].str.contains("Chemnitz")]
chemnitz = ch.sum()
chemnitz_series = pd.Series(chemnitz)
chemnitz_series["WK-Nr"] = 1012
chemnitz_series["WK-Name"] = "Chemnitz 1-3"
chemnitz_series["AGS"] = 14511000
df_lw_2014 = df_lw_2014[~df_lw_2014["Ortname"].str.contains("Chemnitz")]

In [115]:
# Städte-Series an Dataframe anhängen
df_lw_2014 = df_lw_2014.append([leipzig_series, dresden_series, zwickau_series, chemnitz_series], ignore_index=True)

In [116]:
test = df_lw_2014[df_lw_2014["Ortname"].str.contains("Leipzig")]
test

Wahl  WK-Nr      WK-Name           Ebene  \
427  LW14LW14LW14LW14LW14LW14LW14   2733  Leipzig 1-7  TGTGTGTGTGTGTG   

          AGS                                            Ortname  \
427  14713000  Leipzig 1Leipzig 2Leipzig 3Leipzig 4Leipzig 5L...   

    Briefwahl_Sonderfall  Wahlberechtigte  Wähler  ungültige_1      ...       \
427                    0           432324  191422         2201      ...        

     GRÜNE_2  NPD_2  Tierschutzpartei_2  PIRATEN_2  BüSo_2  DSU_2  AfD_2  \
427    21342   6331                2845       3454     440    223  13862   

     pro Deutschland_2  FREIE WÄHLER_2  Die PARTEI_2  
427                269            1824          2534  

[1 rows x 44 columns]

In [117]:
# Löschen von überflüssigen Columns 
df_lw_2014.drop(columns=['Wahl','Ebene', 'Ortname', 'Briefwahl_Sonderfall', 'ungültige_1', 'ungültige_2'], inplace=True) 

In [118]:
df_lw_2014.to_csv("./Rohdaten/Testdaten_aus_LTW_2014.csv", encoding="utf-8", sep=";")

In [119]:
# Umbennen der Spalten 
df_lw_2014 = df_lw_2014.add_suffix("_LW2014")

## Vorbereitung Bundestagswahl 2017

In [120]:
df_bw_2017 = pd.read_excel("./Rohdaten/BW_2017_original.xlsx", sheet_name="BW17_Ergebnisse_GE_TG")

In [121]:
#Für Ausgangsgrafik für Dresden und Leipzig Wahlkreise zusammenziehen

le = df_bw_2017[df_bw_2017["Ortname"].str.contains("Leipzig")]
leipzig = le.sum()
leipzig_series_bt = pd.Series(leipzig)
leipzig_series_bt["WK-Nr"] = 152153
leipzig_series_bt["WK-Name"] = "Leipzig 1-2"
leipzig_series_bt["AGS"] = 14713000

dr = df_bw_2017[df_bw_2017["Ortname"].str.contains("Dresden")]
dresden = dr.sum()
dresden_series_bt = pd.Series(dresden)
dresden_series_bt["WK-Nr"] = 152153
dresden_series_bt["WK-Name"] = "Leipzig 1-2"
dresden_series_bt["AGS"] = 14713000
#Lösche die einzelnen Wahlkreise aus der Datei, da diese ansonsten falsche Gesamtzahlen aufweist... 
df_bw_2017 = df_bw_2017[~df_bw_2017["Ortname"].str.contains("Leipzig")]
df_bw_2017 = df_bw_2017[~df_bw_2017["Ortname"].str.contains("Dresden")]
df_bw_2017 = df_bw_2017.append([leipzig_series_bt, dresden_series_bt], ignore_index=True)

C:\Users\Felix\Anaconda3\lib\site-packages\pandas\core\indexing.py:1472: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
https://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
  return self._getitem_tuple(key)


In [122]:
# Löschen von überflüssigen Columns 
df_bw_2017.drop(columns=['Wahl', 'WK-Nr', 'WK-Name','Ebene', 'Ortname', 'Briefwahl_Sonderfall', \
                         'ungültige_1', 'ungültige_2'], inplace=True) 
# hier können zudem alle Erststimmen-Ergebnisse gelöscht werden, brauchen nur die historischen Zweitstimmenergebnisse
df_bw_2017.drop(columns=['gültige_1', 'CDU_1', 'DIE LINKE_1', 'SPD_1', 'AfD_1', 'GRÜNE_1',
       'NPD_1', 'FDP_1', 'PIRATEN_1', 'FREIE WÄHLER_1', 'BüSo_1', 'MLPD_1',
       'ÖDP_1', 'Die PARTEI_1', 'Bürgerkandidat_1', 'Der Auserwählte_1',
       'Heimat Region Zwickau_1', 'Held_1', 'Hoyer_1',
       'Internationalistisch. Bündnis_1', 'RINK_1', 'SGP_1',
       'Wir sind LEIPZIGER_1'], inplace=True) 
# jetzt löschen wir noch alle Zweitstimmenergebnisse bis auf unsere sieben stärksten Parteien
df_bw_2017.drop(columns=['NPD_2', 'PIRATEN_2', 'BüSo_2', 'MLPD_2', 'BGE_2', 'DiB_2','ÖDP_2', \
                         'Die PARTEI_2', 'Tierschutzpartei_2', 'V-Partei³_2'], inplace=True)

# Endung an alle Spalten hängen
df_bw_2017 = df_bw_2017.add_suffix("_BW2017")

In [123]:
# check ob alle Zahlen int64 sind
df_bw_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 422 entries, 0 to 421
Data columns (total 11 columns):
AGS_BW2017                422 non-null int64
Wahlberechtigte_BW2017    422 non-null int64
Wähler_BW2017             422 non-null int64
gültige_2_BW2017          422 non-null int64
CDU_2_BW2017              422 non-null int64
DIE LINKE_2_BW2017        422 non-null int64
SPD_2_BW2017              422 non-null int64
AfD_2_BW2017              422 non-null int64
GRÜNE_2_BW2017            422 non-null int64
FDP_2_BW2017              422 non-null int64
FREIE WÄHLER_2_BW2017     422 non-null int64
dtypes: int64(11)
memory usage: 36.3 KB


## Vorbereitung Europawahl 2019

In [124]:
df_ew_2019 = pd.read_excel("./Rohdaten/EW_2019_original.xlsx", sheet_name="EW19_endgErgebnisse_GE&KS")

In [125]:
df_ew_2019.columns

Index(['Wahl', 'WK-Nr', 'WK-Name', 'Ebene', 'AGS', 'Ortname',
       'Briefwahl_Sonderfall', 'genaue Briefwahlzuordnung', 'A1', 'A2',
       ...
       'DIE FRAUEN in %', 'Graue Panther in %', 'LKR in %',
       'MENSCHLICHE WELT in %', 'NL in %', 'ÖkoLinX in %',
       'Die Humanisten in %', 'PARTEI FÜR DIE TIERE in %',
       'Gesundheitsforschung in %', 'Volt in %'],
      dtype='object', length=101)

In [126]:
df_ew_2019.drop(columns=['Wahl', 'WK-Nr', 'WK-Name','Ebene', 'Ortname', 'Briefwahl_Sonderfall', 'A1', 'A2', 'A3',\
                          'genaue Briefwahlzuordnung', 'B1','ungültige'], inplace=True) 

# jetzt löschen wir noch alle Zweitstimmenergebnisse bis auf unsere sieben stärksten Parteien
df_ew_2019.drop(columns=['DIE FRAUEN in %', 'Graue Panther in %', 'LKR in %',
       'MENSCHLICHE WELT in %', 'NL in %', 'ÖkoLinX in %',
       'Die Humanisten in %', 'PARTEI FÜR DIE TIERE in %',
       'Gesundheitsforschung in %', 'Volt in %', 'NPD', 
                        'PIRATEN',
       'Tierschutzpartei', 'FAMILIE', 'Die PARTEI', 'Volksabstimmung', 'ÖDP',
       'DKP', 'MLPD', 'BP', 'SGP', 'TIERSCHUTZ hier!', 'Tierschutzallianz',
       'Bündnis C', 'BIG', 'BGE', 'DIE DIREKTE!', 'DiEM25', 'III. Weg',
       'Die Grauen', 'DIE RECHTE', 'DIE VIOLETTEN', 'LIEBE', 'DIE FRAUEN',
       'Graue Panther', 'LKR', 'MENSCHLICHE WELT', 'NL', 'ÖkoLinX',
       'Die Humanisten', 'PARTEI FÜR DIE TIERE', 'Gesundheitsforschung',
       'Volt', 'Unnamed: 56', 'Wahlbeteiligung', 'B1 in %', 'ungültige in %',
       'gültige in %', 'CDU in %', 'DIE LINKE in %', 'SPD in %', 'AfD in %',
       'GRÜNE in %', 'NPD in %', 'FDP in %', 'FREIE WÄHLER in %',
       'PIRATEN in %', 'Tierschutzpartei in %', 'FAMILIE in %',
       'Die PARTEI in %', 'Volksabstimmung in %', 'ÖDP in %', 'DKP in %',
       'MLPD in %', 'BP in %', 'SGP in %', 'TIERSCHUTZ hier! in %',
       'Tierschutzallianz in %', 'Bündnis C in %', 'BIG in %', 'BGE in %',
       'DIE DIREKTE! in %', 'DiEM25 in %', 'III. Weg in %', 'Die Grauen in %',
       'DIE RECHTE in %', 'DIE VIOLETTEN in %', 'LIEBE in %',
       'DIE FRAUEN in %', 'Graue Panther in %', 'LKR in %',
       'MENSCHLICHE WELT in %', 'NL in %', 'ÖkoLinX in %',
       'Die Humanisten in %', 'PARTEI FÜR DIE TIERE in %',
       'Gesundheitsforschung in %', 'Volt in %'], inplace=True)

# Endung an alle Spalten hängen
df_ew_2019 = df_ew_2019.add_suffix("_EW2019")

In [127]:
# check ob alle Zahlen int64 sind
df_ew_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 419 entries, 0 to 418
Data columns (total 11 columns):
AGS_EW2019                419 non-null int64
Wahlberechtigte_EW2019    419 non-null int64
Wähler_EW2019             419 non-null int64
gültige_EW2019            419 non-null int64
CDU_EW2019                419 non-null int64
DIE LINKE_EW2019          419 non-null int64
SPD_EW2019                419 non-null int64
AfD_EW2019                419 non-null int64
GRÜNE_EW2019              419 non-null int64
FDP_EW2019                419 non-null int64
FREIE WÄHLER_EW2019       419 non-null int64
dtypes: int64(11)
memory usage: 36.1 KB


## Zusammenführen der historischen Daten 
## auf der AGS-Nummer

**Problem: ** Es gab noch Gemeindezusammenlegungen. 2013 noch 446 Kommunen, 2017 424, 2019 419. 
** Lösung:  ** Ich joine einfach von links (aktuell), dann werden nur die Geodaten angezeigt, die aktuell vorhanden sind. 
Und: Wenn es keine Vergleichsdaten gibt, kann Javascript später einfach eine Erklärung auswerfen. 


2018: Aue + Bad Schlema = Aue-Bad Schlema



In [128]:
df_wahlen_historisch = df_ew_2019.merge(df_bw_2017, how="left", left_on="AGS_EW2019", right_on="AGS_BW2017").fillna(0)

In [129]:
df_wahlen_historisch = df_wahlen_historisch.merge(df_lw_2014, how="left", left_on="AGS_EW2019", right_on="AGS_LW2014")

In [130]:
# AGS-Nummern wieder löschen bzw. AGS_EW2019 umbenennen
df_wahlen_historisch.drop(columns=["AGS_BW2017", "AGS_LW2014"], inplace=True)
df_wahlen_historisch.rename(columns={"AGS_EW2019":"AGS"}, inplace=True)

In [134]:
df_wahlen_historisch.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 420 entries, 0 to 419
Data columns (total 58 columns):
AGS                              420 non-null int64
Wahlberechtigte_EW2019           420 non-null int64
Wähler_EW2019                    420 non-null int64
gültige_EW2019                   420 non-null int64
CDU_EW2019                       420 non-null int64
DIE LINKE_EW2019                 420 non-null int64
SPD_EW2019                       420 non-null int64
AfD_EW2019                       420 non-null int64
GRÜNE_EW2019                     420 non-null int64
FDP_EW2019                       420 non-null int64
FREIE WÄHLER_EW2019              420 non-null int64
Wahlberechtigte_BW2017           420 non-null int64
Wähler_BW2017                    420 non-null int64
gültige_2_BW2017                 420 non-null int64
CDU_2_BW2017                     420 non-null int64
DIE LINKE_2_BW2017               420 non-null int64
SPD_2_BW2017                     420 non-null int64
AfD_2_BW201

In [132]:
# richtige Datentypen vergeben - alle Zahlen np.int64
cols = ['AGS', 'Wahlberechtigte_EW2019', 'Wähler_EW2019', 'gültige_EW2019',
       'CDU_EW2019', 'DIE LINKE_EW2019', 'SPD_EW2019', 'AfD_EW2019',
       'GRÜNE_EW2019', 'FDP_EW2019', 'FREIE WÄHLER_EW2019',
       'Wahlberechtigte_BW2017', 'Wähler_BW2017', 'gültige_2_BW2017',
       'CDU_2_BW2017', 'DIE LINKE_2_BW2017', 'SPD_2_BW2017', 'AfD_2_BW2017',
       'GRÜNE_2_BW2017', 'FDP_2_BW2017', 'FREIE WÄHLER_2_BW2017',
       'WK-Nr_LW2014','Wahlberechtigte_LW2014',
       'Wähler_LW2014', 'gültige_1_LW2014', 'CDU_1_LW2014',
       'DIE LINKE_1_LW2014', 'SPD_1_LW2014', 'FDP_1_LW2014', 'GRÜNE_1_LW2014',
       'NPD_1_LW2014', 'PIRATEN_1_LW2014', 'BüSo_1_LW2014', 'DSU_1_LW2014',
       'AfD_1_LW2014', 'FREIE WÄHLER_1_LW2014', 'Die PARTEI_1_LW2014',
       'Freie Bürger Leipzig_1_LW2014', 'Konrad Skatula_1_LW2014',
       'Schaar_1_LW2014', 'WAGNER_1_LW2014', 'WV MFU_1_LW2014',
       'gültige_2_LW2014', 'CDU_2_LW2014', 'DIE LINKE_2_LW2014',
       'SPD_2_LW2014', 'FDP_2_LW2014', 'GRÜNE_2_LW2014', 'NPD_2_LW2014',
       'Tierschutzpartei_2_LW2014', 'PIRATEN_2_LW2014', 'BüSo_2_LW2014',
       'DSU_2_LW2014', 'AfD_2_LW2014', 'pro Deutschland_2_LW2014',
       'FREIE WÄHLER_2_LW2014', 'Die PARTEI_2_LW2014']

df_wahlen_historisch[cols] = df_wahlen_historisch[cols].fillna(0).astype(np.int64)

In [133]:
df_wahlen_historisch.to_csv("./Daten_clean/historische_Wahldaten_bereinigt.csv", sep=";", encoding="utf-8")